### ISSUE TITLE
Horovod does not work well with Adagrad pytorch

### ISSUE THREAD

**Issue by adilkhanovkz** [uber/horovod#697]
When I try to use horovod.DistributedOptimizer with Adagrad, loss reaches infinity, while with the SGD it works correctly. What is wrong? 

    optimizer = optim.Adagrad(model.parameters(), 
                          lr=0.15 *float(batch_size)/16. * hvd.size(),
                          initial_accumulator_value=0.1)

    compression = hvd.Compression.fp16
    optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters(), 
    compression=compression)

    hvd.broadcast_parameters(model.state_dict(), root_rank=0)
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)

    for epoch in range(num_epochs):    
        for batch_idx, (batch_contexts,
                    batch_contexts_oov,
                    batch_highlights_in,
                    batch_highlights_out,
                    batch_oov_vocab
                   ) in enumerate(train_loader):

            [batch_contexts, \
            batch_contexts_oov, \
            batch_highlights_in, \
            batch_highlights_out] = \
                                    [i.to(device) for i in [batch_contexts, \
                                                    batch_contexts_oov, \
                                                    batch_highlights_in, \
                                                    batch_highlights_out]]
        
            optimizer.zero_grad()
            probs, coverage = model(batch_contexts,  batch_contexts_oov, batch_highlights_in)
        
            probs = probs.view(-1,probs.size(-1))
            batch_highlights_out = batch_highlights_out.view(-1)
            mask = batch_highlights_out != data.words_vocab.pad_idx
            loss = F.nll_loss(probs[mask], batch_highlights_out[mask])
        
            loss.backward()
            optimizer.synchronize()
            torch.nn.utils.clip_grad_norm(model.parameters(), 2)
            optimizer.step()

**Comment by tgaddair** [comment]
Hey @adilkhanovkz, does this problem occur without fp16 compression?

**Comment by adilkhanovkz** [comment]
Hello @tgaddair! Thank you for reply! Yes, this problem occur without fp16 compression too. I tried different combinations, but it does not help. Without hvd.DistributedOptimizer  (without horovod in one node) same code works better with Adagrad rather than with SGD. I did not find examples of using horovod with Adagrad. Probably I missing something.(

**Comment by adilkhanovkz** [comment]
next test:
   import torch.optim as optim
   import torch.nn as nn
   import horovod.torch as hvd

   class Model(nn.Module):
       def __init__(self):
           super(Model,self).__init__()
           self.linear = nn.Linear(10,10)
       def forward(self,x):
           return self.linear(x)

   model = Model()
   optimizer = optim.Adagrad(model.parameters(), lr=0.15,initial_accumulator_value=0.1)
   print('Adagrad initial defaults: ', optimizer.defaults)
   optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
   print('Adagrad distrib defaults: ', optimizer.defaults)

   optimizer = optim.SGD(model.parameters(), lr=0.15,momentum=0.1)
   print('SGD initial defaults: ', optimizer.defaults)
   optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
   print('SGD distrib defaults: ', optimizer.defaults)

shows nex output:
   Adagrad initial defaults:  {'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}
   Adagrad distrib defaults:  {'lr': 0.01, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0}
   SGD initial defaults:  {'lr': 0.15, 'momentum': 0.1, 'dampening': 0, 'weight_decay': 0, 'nesterov': False}
   SGD distrib defaults:  {'lr': <required parameter>, 'momentum': 0, 'dampening': 0, 'weight_decay': 0, 'nesterov': False}

**Comment by adilkhanovkz** [comment]
Actually initial_accumulator_value does not inherited in hvd.DistributedOptimizer

**Comment by adilkhanovkz** [comment]
import torch.optim as optim
import torch.nn as nn
import horovod.torch as hvd
from horovod.torch.compression import Compression
import torch
hvd.init()

class Model(nn.Module):
    def __init__(self):
        super(Model,self).__init__()
        self.linear = nn.Linear(5,2)
    def forward(self,x):
        return self.linear(x)

model1 = Model()
model2 = Model()

model1.load_state_dict(model2.state_dict())

criterion = nn.NLLLoss()

optimizer1 = optim.Adagrad(model1.parameters(), lr=0.15,initial_accumulator_value=0.1)
print('\nAdagrad initial optimizer1 defaults: ', optimizer1.defaults)
print('\nAdagrad initial optimizer1 params: ', optimizer1.param_groups)

optimizer2 = optim.Adagrad(model2.parameters(), lr=0.15,initial_accumulator_value=0.1)
print('\nAdagrad initial optimizer2 defaults: ', optimizer2.defaults)
print('\nAdagrad initial optimizer2 params: ', optimizer2.param_groups)

optimizer2 = hvd.DistributedOptimizer(optimizer2, named_parameters=model2.named_parameters())
print('\nAdagrad distrib optimizer2 defaults: ', optimizer2.defaults)
print('\nAdagrad distrib optimizer2 params: ', optimizer2.param_groups)
```

    
    Adagrad initial optimizer1 defaults:  {'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}
    
    Adagrad initial optimizer1 params:  [{'params': [Parameter containing:
    tensor([[ 0.1313, -0.3276, -0.3194, -0.2915, -0.2523],
            [-0.2636, -0.0318,  0.1258, -0.1886, -0.2429]], requires_grad=True), Parameter containing:
    tensor([0.0443, 0.1068], requires_grad=True)], 'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}]
    
    Adagrad initial optimizer2 defaults:  {'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}
    
    Adagrad initial optimizer2 params:  [{'params': [Parameter containing:
    tensor([[ 0.1313, -0.3276, -0.3194, -0.2915, -0.2523],
            [-0.2636, -0.0318,  0.1258, -0.1886, -0.2429]], requires_grad=True), Parameter containing:
    tensor([0.0443, 0.1068], requires_grad=True)], 'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}]
    
    Adagrad distrib optimizer2 defaults:  {'lr': 0.01, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0}
    
    Adagrad distrib optimizer2 params:  [{'params': [Parameter containing:
    tensor([[ 0.1313, -0.3276, -0.3194, -0.2915, -0.2523],
            [-0.2636, -0.0318,  0.1258, -0.1886, -0.2429]], requires_grad=True), Parameter containing:
    tensor([0.0443, 0.1068], requires_grad=True)], 'lr': 0.15, 'lr_decay': 0, 'weight_decay': 0, 'initial_accumulator_value': 0.1}]



```python
optimizer1.zero_grad()
optimizer2.zero_grad()
```


```python
optimizer1.state_dict()
```




    {'state': {139936924435584: {'step': 0,
       'sum': tensor([[0.1000, 0.1000, 0.1000, 0.1000, 0.1000],
               [0.1000, 0.1000, 0.1000, 0.1000, 0.1000]])},
      139936924435656: {'step': 0, 'sum': tensor([0.1000, 0.1000])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435584, 139936924435656]}]}




```python
optimizer2.state_dict()
```




    {'state': {139936924435728: {'step': 0, 'sum': tensor([[0., 0., 0., 0., 0.],
               [0., 0., 0., 0., 0.]])},
      139936924435800: {'step': 0, 'sum': tensor([0., 0.])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435728, 139936924435800]}]}




```python
optimizer2.load_state_dict(optimizer1.state_dict()) ### this helps
```


```python
optimizer1.state_dict()
```




    {'state': {139936924435584: {'step': 0,
       'sum': tensor([[0.1000, 0.1000, 0.1000, 0.1000, 0.1000],
               [0.1000, 0.1000, 0.1000, 0.1000, 0.1000]])},
      139936924435656: {'step': 0, 'sum': tensor([0.1000, 0.1000])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435584, 139936924435656]}]}




```python
optimizer2.state_dict()
```




    {'state': {139936924435728: {'step': 0,
       'sum': tensor([[0.1000, 0.1000, 0.1000, 0.1000, 0.1000],
               [0.1000, 0.1000, 0.1000, 0.1000, 0.1000]])},
      139936924435800: {'step': 0, 'sum': tensor([0.1000, 0.1000])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435728, 139936924435800]}]}




```python
list(model1.parameters())
```




    [Parameter containing:
     tensor([[ 0.1313, -0.3276, -0.3194, -0.2915, -0.2523],
             [-0.2636, -0.0318,  0.1258, -0.1886, -0.2429]], requires_grad=True),
     Parameter containing:
     tensor([0.0443, 0.1068], requires_grad=True)]




```python
list(model2.parameters())
```




    [Parameter containing:
     tensor([[ 0.1313, -0.3276, -0.3194, -0.2915, -0.2523],
             [-0.2636, -0.0318,  0.1258, -0.1886, -0.2429]], requires_grad=True),
     Parameter containing:
     tensor([0.0443, 0.1068], requires_grad=True)]




```python
input = torch.randn(3, 5)
```


```python
output1 = model1(input)
output2 = model2(input)
```


```python
loss1 = criterion(output1,torch.tensor([1, 0,0]))
loss2 = criterion(output2,torch.tensor([1, 0,0]))
```


```python
loss1.backward()
loss2.backward()
```


```python
optimizer1.step()
optimizer2.step()
```


```python
optimizer1.state_dict()
```




    {'state': {139936924435584: {'step': 1,
       'sum': tensor([[0.7409, 0.9909, 0.3507, 0.3140, 0.4023],
               [0.1767, 0.1124, 0.1996, 0.3744, 0.2627]])},
      139936924435656: {'step': 1, 'sum': tensor([0.5444, 0.2111])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435584, 139936924435656]}]}




```python
optimizer2.state_dict()
```




    {'state': {139936924435728: {'step': 1,
       'sum': tensor([[0.7409, 0.9909, 0.3507, 0.3140, 0.4023],
               [0.1767, 0.1124, 0.1996, 0.3744, 0.2627]])},
      139936924435800: {'step': 1, 'sum': tensor([0.5444, 0.2111])}},
     'param_groups': [{'lr': 0.15,
       'lr_decay': 0,
       'weight_decay': 0,
       'initial_accumulator_value': 0.1,
       'params': [139936924435728, 139936924435800]}]}

**Comment by adilkhanovkz** [comment]
I find next temporary solution:
    optimizer2.load_state_dict(optimizer1.state_dict())
